src/commands/sync/sync.ts

```diff
@@ -0,0 +1,138 @@
+import type { BaseOptions } from '../../options';
+import type Client from '../../client';
+import { Command } from 'commander';
+import ansi from 'ansi-colors';
+
+import { extractKeysOfFiles, filterExtractionResult } from '../../extractor';
+import { dumpWarnings } from '../../extractor/warnings';
+import { compareKeys } from './comparator';
+
+import { overwriteDir } from '../../utils/overwriteDir';
+import { unzipBuffer } from '../../utils/zip';
+import { loading, error } from '../../utils/logger';
+
+import { EXTRACTOR } from '../../options';
+
+type Options = BaseOptions & {
+  extractor: string;
+  backup?: string;
+  keepUnused?: boolean;
+  abortOnWarning?: boolean;
+};
+
+async function backup(client: Client, dest: string) {
+  const blob = await client.export.export({
+    format: 'JSON',
+    filterState: ['UNTRANSLATED', 'TRANSLATED', 'REVIEWED'],
+    structureDelimiter: '',
+  });
+
+  await unzipBuffer(blob, dest);
+}
+
+async function syncHandler(this: Command, pattern: string) {
+  const opts: Options = this.optsWithGlobals();
+
+  const rawKeys = await extractKeysOfFiles(pattern, opts.extractor);
+  const warnCount = dumpWarnings(rawKeys);
+  if (opts.abortOnWarning && warnCount) {
+    console.log(ansi.bold.red('Aborting as warnings have been emitted.'));
+    process.exit(1);
+  }
+
+  const localKeys = await filterExtractionResult(rawKeys);
+  const remoteKeys = await opts.client.project.fetchAllKeys();
+
+  const diff = compareKeys(localKeys, remoteKeys);
+  if (!diff.added.length && !diff.removed.length) {
+    console.log(
+      ansi.green(
+        'Your code project is in sync with the associated Tolgee project!'
+      )
+    );
+    process.exit(0);
+  }
+
+  // Load project settings. We're interested in the default locale here.
+  const project = await opts.client.project.fetchProjectInformation();
+  if (!project.baseLanguage) {
+    // I'm highly unsure how we could reach this state, but this is what the OAI spec tells me ¯\_(ツ)_/¯
+    error('Your project does not have a base language!');
+    process.exit(1);
+  }
+
+  // Prepare backup
+  if (opts.backup) {
+    await overwriteDir(opts.backup);
+    await loading(
+      'Backing up Tolgee project',
+      backup(opts.client, opts.backup)
+    );
+  }
+
+  // Create new keys
+  if (diff.added.length) {
+    for (const key of diff.added) {
+      await opts.client.project.createKey({
+        name: key.keyName,
+        namespace: key.namespace,
+        translations: key.defaultValue
+          ? { [project.baseLanguage.tag]: key.defaultValue }
+          : undefined,
+      });
+    }
+  }
+
+  if (!opts.keepUnused) {
```

Comment:
Manual confirmation (or `--yes`) has been added in 344b8d6. For a more efficient way to add keys, I'd love to use a single import but the problem of keys without translations makes it impossible.

Maybe this behavior could be changed when adding files on the server, in which case we'd *actually* have a sorta efficient way of adding keys in bulk across multiple namespaces. :D

Reply:
Also, maybe it would be cool to be able to ignore keys with several tag in the platform. Or maybe there should be other way how store keys, which the user doesn't want to delete or add.

Reply:
> Also, maybe it would be cool to be able to ignore keys with several tag in the platform. Or maybe there should be other way how store keys, which the user doesn't want to delete or add.

Opened an [issue](https://github.com/tolgee/tolgee-cli/issues/21) about it. 👍🏻